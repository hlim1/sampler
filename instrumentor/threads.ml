open Cil


let threads =
  Options.registerBoolean
    ~flag:"threads"
    ~desc:"create thread-safe code"
    ~ident:"Threads"
    ~default:false


let bump file =
  if !threads then
    let bumper = Lval (var (FindFunction.find "atomicIncrementCounter" file)) in
    fun counter location ->
      Call (None, bumper, [mkAddrOf counter], location)
  else
    fun counter location ->
      Set (counter, increm (Lval counter) 1, location)
