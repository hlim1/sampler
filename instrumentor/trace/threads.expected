<sites unit="0e501eb35852c536411dd498fda02e89" scheme="data">
instrumentor/trace/threads.c	11	drop_x	4	x	read
instrumentor/trace/threads.c	11	drop_x	7	x	read
instrumentor/trace/threads.c	11	drop_x	10	x	write
instrumentor/trace/threads.c	13	drop_x	21	x	read
instrumentor/trace/threads.c	14	drop_x	26	x	read
instrumentor/trace/threads.c	22	drop_y	4	y	read
instrumentor/trace/threads.c	22	drop_y	7	y	read
instrumentor/trace/threads.c	22	drop_y	10	y	write
instrumentor/trace/threads.c	24	drop_y	21	y	read
instrumentor/trace/threads.c	25	drop_y	26	y	read
instrumentor/trace/threads.c	39	main	5	thread_x	read
instrumentor/trace/threads.c	40	main	9	thread_y	read
</sites>
<sites unit="0e501eb35852c536411dd498fda02e89" scheme="branches">
instrumentor/trace/threads.c	11	drop_x	14	tmp
instrumentor/trace/threads.c	14	drop_x	30	cbi_shared_access4 % 10 == 0
instrumentor/trace/threads.c	22	drop_y	14	tmp
instrumentor/trace/threads.c	25	drop_y	30	cbi_shared_access4 % 5 == 0
</sites>
<sites unit="0e501eb35852c536411dd498fda02e89" scheme="function-entries">
instrumentor/trace/threads.c	9	drop_x	0
instrumentor/trace/threads.c	20	drop_y	0
instrumentor/trace/threads.c	31	main	0
</sites>
